onnxruntime/core/providers/cuda/math/topk.cc

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "topk.h"
+#include "topk_impl.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+ONNX_OPERATOR_VERSIONED_KERNEL_EX(
+    TopK,
+    kOnnxDomain,
+    1,9,
+    kCudaExecutionProvider,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllFixedSizeTensorTypes()),
+    TopK<false>);
+
+ONNX_OPERATOR_VERSIONED_KERNEL_EX(
+    TopK,
+    kOnnxDomain,
+    10,11,
+    kCudaExecutionProvider,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::AllFixedSizeTensorTypes()),
+    TopK<true>);
+
+template <bool inputk>
+TopK<inputk>::TopK(const OpKernelInfo& info) : CudaKernel(info) {
+  info.GetAttrOrDefault<int64_t>("axis", &axis_, -1);
+  info.GetAttrOrDefault<int64_t>("largest", &largest_, 1);
+  info.GetAttrOrDefault<int64_t>("sorted", &sorted_, 1);
+  if (!inputk) {
+    info.GetAttrOrDefault<int64_t>("k", &K_, 0);
+  }
+}
+
+#define ISTYPE(T) tensor_X->DataType() == DataTypeImpl::GetType<T>()
+#define TOPKIMPL(T) TopKImpl<T>(tensor_X->Data<T>(),                               \
+                                static_cast<T*>(tensor_V->MutableDataRaw()),       \
+                                static_cast<int64_t*>(tensor_I->MutableDataRaw()), \
+                                elem_nums_cuda.GpuPtr(),                           \
+                                elem_nums.size(),                                  \
+                                axis, K_, largest_, sorted_, N, dimension)
+
+template <bool inputk>
+Status TopK<inputk>::ComputeInternal(OpKernelContext* ctx) const {
+  auto tensor_X = ctx->Input<Tensor>(0);
+  ORT_ENFORCE(nullptr != tensor_X);
+  auto rank = static_cast<int64_t>(tensor_X->Shape().NumDimensions());
+  auto axis = axis_ < 0 ? rank + axis_ : axis_;
+  ORT_ENFORCE(axis > -1 && axis < rank);
+
+  if (inputk) {
+    auto tensor_K = ctx->Input<Tensor>(1);
+    ORT_ENFORCE(nullptr != tensor_K);
+    cudaMemcpyAsync(&K_, tensor_K->Data<int64_t>(), sizeof(int64_t), cudaMemcpyDeviceToHost);
```

Comment:
>cudaMemcpyAsync(&K_, tensor_K->Data<int64_t>(), sizeof(int64_t), cudaMemcpyDeviceToHost); [](start = 4, length = 89)

cudaMemcpyAsync is async with CPU. It will return immediately.

Reply:
done.